crypto/cmp/cmp_ctx.c

```diff
@@ -20,43 +20,49 @@
 #include <openssl/crmf.h>
 #include <openssl/err.h>
 
+#define DEFINE_OSSL_CMP_CTX_get0(FIELD, TYPE) \
+    DEFINE_OSSL_CMP_CTX_get0_NAME(FIELD, FIELD, TYPE)
+#define DEFINE_OSSL_CMP_CTX_get0_NAME(NAME, FIELD, TYPE) \
+TYPE *OSSL_CMP_CTX_get0_##NAME(const OSSL_CMP_CTX *ctx) \
+{ \
+    if (ctx == NULL) { \
+        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT); \
+        return NULL; \
+    } \
+    return ctx->FIELD; \
+}
+
 /*
  * Get current certificate store containing trusted root CA certs
  */
-X509_STORE *OSSL_CMP_CTX_get0_trustedStore(const OSSL_CMP_CTX *ctx)
-{
-    if (ctx == NULL) {
-        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT);
-        return NULL;
-    }
-    return ctx->trusted;
+DEFINE_OSSL_CMP_CTX_get0_NAME(trustedStore, trusted, X509_STORE)
+
+#define DEFINE_OSSL_set0(PREFIX, FIELD, TYPE) \
+    DEFINE_OSSL_set0_NAME(PREFIX, FIELD, FIELD, TYPE)
+#define DEFINE_OSSL_set0_NAME(PREFIX, NAME, FIELD, TYPE) \
+int PREFIX##_set0##_##NAME(OSSL_CMP_CTX *ctx, TYPE *val) \
+{ \
+    if (ctx == NULL) { \
+        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT); \
+        return 0; \
+    } \
+    TYPE##_free(ctx->FIELD); \
+    ctx->FIELD = val; \
+    return 1; \
 }
 
 /*
  * Set certificate store containing trusted (root) CA certs and possibly CRLs
  * and a cert verification callback function used for CMP server authentication.
  * Any already existing store entry is freed. Given NULL, the entry is reset.
  */
-int OSSL_CMP_CTX_set0_trustedStore(OSSL_CMP_CTX *ctx, X509_STORE *store)
-{
-    if (ctx == NULL) {
-        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT);
-        return 0;
-    }
-    X509_STORE_free(ctx->trusted);
-    ctx->trusted = store;
-    return 1;
-}
+DEFINE_OSSL_set0_NAME(OSSL_CMP_CTX, trustedStore, trusted, X509_STORE)
 
 /* Get current list of non-trusted intermediate certs */
-STACK_OF(X509) *OSSL_CMP_CTX_get0_untrusted(const OSSL_CMP_CTX *ctx)
-{
-    if (ctx == NULL) {
-        ERR_raise(ERR_LIB_CMP, CMP_R_NULL_ARGUMENT);
-        return NULL;
-    }
-    return ctx->untrusted;
-}
+DEFINE_OSSL_CMP_CTX_get0(untrusted, STACK_OF(X509))
+
+#define X509_STACK_free(certs) \
```

Comment:
I agree that a function would even be better here (while if it's just used in this file, I would not bother changing it this way).
I even suggest making it a public function because deallocating a stack of certs is pretty common, and it is rather cumbersome to write (or copy&paste) `sk_X509_pop_free(cert, X509_free)` each time.

If you agree,  shall I add a commit to this PR or better set up a new one,
and which name should we choose for the public function?

Reply:
It needn't be a public function, just not local to the file: `ossl_x509_stack_free()` e.g.
The question would be how widely used would it be outside of OpenSSL?

Reply:
In the libraries that I build on OpenSSL, I had this pattern sufficiently often to create an aux function for it.
Not sure if others here would see it as helpful, too?

Reply:
That would tend to indicate making it public.

Reply:
> That would tend to indicate making it public.

Fine. Under which name?

Reply:
Would `X509_STACK_free()` be acceptable?
Or the more formal `STACK_OF_X509_free()?

Would it be better to merge this PR first, and then take some time to come to a conclusion if and how to introduce the function, as there is no rush on this?

Reply:
I'd be fine with `X509_STACK_free` but it might be better to be `OSSL_STACK_OF_X509_free` which adheres to our public symbol naming convention.

Reply:
I see - so I'm goint do that, but in a separate PR since this was is already approved and ready and I would use a separate commit anyway.

Reply:
Added respective PR: #17307.